File: SparkStreamingDemo/src/main/java/cn/edu/ecnu/sparkstreaming/example/java/wordcount/WordCount.java
package cn.edu.ecnu.sparkstreaming.example.java.wordcount;

import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
import org.apache.spark.SparkConf;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import scala.Tuple2;

import java.util.Arrays;

public class WordCount {

    public static void main(String[] args) throws Exception {
        // 1 创建 JavaStreamingContext
        SparkConf sparkConf = new SparkConf();
        sparkConf.setAppName("WordCountJava");
//        sparkConf.setMaster("local[2]"); // 仅用于本地调试, 放集群中运行时删除本行
        JavaStreamingContext sc = new JavaStreamingContext(sparkConf, Durations.seconds(10)); // 处理间隔为 10s
        sc.sparkContext().setLogLevel("ERROR");

        // 2 处理数据
        JavaPairDStream<String, Integer> wordCount = sc
                // 接收 socket 流数据, 其中 args[0] 为 IP 地址, args[1] 为端口号
                .socketTextStream(args[0], Integer.parseInt(args[1]))
                // 将接收到的字符串按空格分割
                .flatMap((String line) -> Arrays.asList(line.split(" ")).iterator())
                // 映射单词为 (word, 1)
                .mapToPair((String word) -> new Tuple2<>(word, 1))
                // 窗口操作, 窗口长度是 20s, 滑动时间间隔是 10s, 即每隔 10s 统计前 20s 的单词
                .reduceByKeyAndWindow(
                        (Integer v1, Integer v2) -> v1 + v2,
                        Durations.seconds(20),
                        Durations.seconds(10)
                );

        // 3 显示结果
        wordCount.print();

        // 4 将结果输出到文件中, 其中 args[2] 为输出路径
        wordCount.saveAsNewAPIHadoopFiles(
                args[2] + "/stream", "output", Text.class, IntWritable.class, TextOutputFormat.class
        );

        // 5 开启StreamingContext
        sc.start();
        sc.awaitTermination();
    }

}

File: SparkDemo/src/main/scala/cn/edu/ecnu/spark/example/scala/wordcount/WordCount.scala
package cn.edu.ecnu.spark.example.scala.wordcount

import org.apache.spark.{SparkConf, SparkContext}

object WordCount {

  def run(args: Array[String]): Unit = {
    /* 步骤1：通过SparkConf设置配置信息，并创建SparkContext */
    val conf = new SparkConf
    conf.setAppName("WordCount")
    conf.setMaster("local") // 仅用于本地进行调试，如在集群中运行则删除本行
    val sc = new SparkContext(conf)

    /* 步骤2：按应用逻辑使用操作算子编写DAG，其中包括RDD的创建、转换和行动等 */
    // 读入文本数据，创建名为lines的RDD
    val lines = sc.textFile(args(0))
    // 将lines中的每一个文本行按空格分割成单个单词
    val words = lines.flatMap { line => line.split(" ") }
    // 将每个单词的频数设置为1，即将每个单词映射为[单词, 1]
    val pairs = words.map { word => (word, 1) }

    // 按单词聚合，并对相同单词的频数使用sum进行累计
    val wordCounts = pairs.groupByKey().map(t => (t._1, t._2.sum))
    // 如需使用合并机制则将第上一行替换为下行
    // val wordCounts = pairs.reduceByKey(_+_)

    // 输出词频统计结果到文件
    wordCounts.saveAsTextFile(args(1))

    /* 步骤3：关闭SparkContext */
    sc.stop()
  }

  def main(args: Array[String]): Unit = {
    run(args)
  }
}
